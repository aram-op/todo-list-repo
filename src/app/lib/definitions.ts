export type Todo =  {
    id: string,
    title: string,
    is_completed: boolean
}