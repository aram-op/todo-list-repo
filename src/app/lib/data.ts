import {createPool} from '@vercel/postgres';
import {Todo} from '@/app/lib/definitions';

const pool = createPool({
    connectionString: process.env.NEXT_PUBLIC_POSTGRES_URL,
});

export async function fetchFilteredTodos(query: string) {

    try {
        const todos = await pool.sql<Todo>`
            SELECT * FROM todos WHERE UPPER(title) LIKE UPPER(${`%${query}%`});
        `;

        return todos.rows;
    } catch (error) {
        console.error('Database Error:', error);
        throw new Error('Failed to fetch todos.');
    }
}

export async function fetchTodoById(id: string) {
    try {
        const result = await pool.sql<Todo>`
            SELECT * FROM todos WHERE id = ${id};
        `;

        return result.rows[0];
    } catch (error) {
        console.error('Database Error:', error);
        throw new Error('Failed to fetch todo.');
    }
}

export async function updateTodo(todo: Todo) {
    try {
        const result = await pool.sql`
            UPDATE todos
             SET title = ${`${todo.title}`},
             is_completed = ${`${todo.is_completed}`}
             WHERE id = ${`${todo.id}`}
                RETURNING *;
        `;
        return result.rows[0];
    } catch (error) {
        console.error('Database Error:', error);
        throw new Error('Failed to update todo.');
    }
}

export async function updateTodoTitle(id: string, title: string) {
    try {
        const result = await pool.sql`
            UPDATE todos
             SET title = ${`${title}`},
             WHERE id = ${`${id}`}
                RETURNING *;
        `;
        return result.rows[0];
    } catch (error) {
        console.error('Database Error:', error);
        throw new Error('Failed to update todo.');
    }
}

export async function deleteTodoById(id: string) {
    try {
        const result = await pool.sql`
              DELETE FROM todos WHERE id = ${`${id}`};
        `;
        return result.rows[0];
    } catch (error) {
        console.error('Database Error:', error);
        throw new Error('Failed to delete todo.');
    }
}

export async function addTodo(title: string) {
    try {
        const result = await pool.sql`
            INSERT INTO todos (title, is_completed)
             VALUES
             (${title}, false)
             ON CONFLICT (id) DO NOTHING;
        `;
        return result.rows[0];
    } catch (error) {
        console.error('Database Error:', error);
        throw new Error('Failed to add todo.')
    }
}