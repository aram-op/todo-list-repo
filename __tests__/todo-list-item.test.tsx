import {fireEvent, render, screen, waitFor} from '@testing-library/react';
import {deleteTodoById, updateTodo} from '@/app/lib/data';
import {useRouter} from 'next/navigation';
import {Todo} from '@/app/lib/definitions';
import TodoListItem from '@/app/ui/todo-list-item/todo-list-item';

jest.mock('@/app/lib/data', () => ({
    deleteTodoById: jest.fn(),
    updateTodo: jest.fn(),
}));

jest.mock('next/navigation', () => ({
    useRouter: jest.fn(),
}));

describe('TodoListItem component', () => {
    const mockTodo: Todo = {id: '1', title: 'Go for a walk', is_completed: false};
    const mockOnItemRemoved = jest.fn();
    const mockPush = jest.fn();

    beforeEach(() => {
        jest.clearAllMocks();
        (useRouter as jest.Mock).mockReturnValue({push: mockPush});
    });

    it('should render the todo', () => {
        render(<TodoListItem todo={mockTodo} onItemRemoved={mockOnItemRemoved}/>);

        expect(screen.getByText('Go for a walk')).toBeInTheDocument();
        expect(screen.getByAltText('complete')).toBeInTheDocument();
        expect(screen.getByAltText('remove')).toBeInTheDocument();
    });

    it('should update the is_completed of todo and apply the completed class', async () => {
        render(<TodoListItem todo={mockTodo} onItemRemoved={mockOnItemRemoved}/>);

        const completeButton = screen.getByAltText('complete');
        fireEvent.click(completeButton);

        expect(updateTodo).toHaveBeenCalledWith({...mockTodo, is_completed: true});
        expect(screen.getByText('Go for a walk')).toHaveClass('completed');
    });

    it('should remove the todo and call .onItemRemoved()', async () => {
        render(<TodoListItem todo={mockTodo} onItemRemoved={mockOnItemRemoved}/>);

        const removeButton = screen.getByAltText('remove');
        fireEvent.click(removeButton);

        await waitFor(() => {
            expect(deleteTodoById).toHaveBeenCalledWith(mockTodo.id);
            expect(mockOnItemRemoved).toHaveBeenCalledWith(mockTodo.id);
        });
    });

    it('should navigate to edit page after the user clicked on title', () => {
        render(<TodoListItem todo={mockTodo} onItemRemoved={mockOnItemRemoved}/>);

        const title = screen.getByText('Go for a walk');
        fireEvent.click(title);

        expect(mockPush).toHaveBeenCalledWith(`todos/${mockTodo.id}/edit`);
    });

    it('should match snapshot', () => {
        const {asFragment} = render(<TodoListItem todo={mockTodo} onItemRemoved={mockOnItemRemoved}/>);
        expect(asFragment()).toMatchSnapshot();
    });
});
