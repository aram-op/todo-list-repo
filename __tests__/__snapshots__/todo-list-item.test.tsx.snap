// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoListItem Component matches snapshot 1`] = `
<DocumentFragment>
  <li
    class="item"
  >
    <p
      class="text "
    >
      Sample Todo
    </p>
    <button
      class="complete"
    >
      <img
        alt="complete"
        height="20"
        src="check-mark.svg"
        width="20"
      />
    </button>
    <button
      class="delete"
    >
      <img
        alt="remove"
        height="25"
        src="remove.svg"
        width="25"
      />
    </button>
  </li>
</DocumentFragment>
`;

exports[`TodoListItem component should match snapshot 1`] = `
<DocumentFragment>
  <li
    class="item"
  >
    <p
      class="text "
    >
      Go for a walk
    </p>
    <button
      class="complete"
    >
      <img
        alt="complete"
        height="20"
        src="check-mark.svg"
        width="20"
      />
    </button>
    <button
      class="delete"
    >
      <img
        alt="remove"
        height="25"
        src="remove.svg"
        width="25"
      />
    </button>
  </li>
</DocumentFragment>
`;
